flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala

```diff
@@ -0,0 +1,162 @@
+        "tables as the table would grow infinitely")
+    }
+  }
+
+  override def getRecordType: TypeInformation[Row] = new RowTypeInfo(getFieldTypes, getFieldNames)
+
+  override def emitDataStream(dataStream: DataStream[JTuple2[JBool, Row]]): Unit = {
+    val keyIndices = keys.map(getFieldNames.indexOf(_))
+    val keyTypes = keyIndices.map(getFieldTypes(_))
+
+    val keySelectorType = new RowTypeInfo(keyTypes, keys)
+
+    val processFunction = new QueryableStateProcessFunction(
+      namePrefix,
+      queryConfig,
+      keys,
+      getFieldNames,
+      getFieldTypes)
+
+    dataStream.keyBy(new RowKeySelector(keyIndices, keySelectorType))
+      .process(processFunction)
+  }
+
+  override protected def copy: TableSinkBase[JTuple2[JBool, Row]] = {
+    new QueryableTableSink(this.namePrefix, this.queryConfig)
+  }
+}
+
+class RowKeySelector(
+  private val keyIndices: Array[Int],
+  @transient private val returnType: TypeInformation[Row])
+  extends KeySelector[JTuple2[JBool, Row], Row]
+    with ResultTypeQueryable[Row] {
+
+  override def getKey(value: JTuple2[JBool, Row]): Row = {
+    val keys = keyIndices
+
+    val srcRow = value.f1
+
+    val destRow = new Row(keys.length)
+    var i = 0
+    while (i < keys.length) {
+      destRow.setField(i, srcRow.getField(keys(i)))
+      i += 1
+    }
+
+    destRow
+  }
+
+  override def getProducedType: TypeInformation[Row] = returnType
+}
+
+class QueryableStateProcessFunction(
+  private val namePrefix: String,
+  private val queryConfig: StreamQueryConfig,
+  private val keyNames: Array[String],
+  private val fieldNames: Array[String],
+  private val fieldTypes: Array[TypeInformation[_]])
+  extends ProcessFunctionWithCleanupState[JTuple2[JBool, Row], Void](queryConfig) {
+
+  @transient private var states = Array[ValueState[AnyRef]]()
+  @transient private var nonKeyIndices = Array[Int]()
+
+  override def open(parameters: Configuration): Unit = {
+    super.open(parameters)
+
+    nonKeyIndices = fieldNames.indices
+      .filter(idx => !keyNames.contains(fieldNames(idx)))
+      .toArray
+
+    val statesBuilder = Array.newBuilder[ValueState[AnyRef]]
+
+    for (i <- nonKeyIndices) {
+      val stateDesc = new ValueStateDescriptor(fieldNames(i), fieldTypes(i))
+      stateDesc.initializeSerializerUnlessSet(getRuntimeContext.getExecutionConfig)
+      stateDesc.setQueryable(fieldNames(i))
+      statesBuilder += getRuntimeContext.getState(stateDesc).asInstanceOf[ValueState[AnyRef]]
+    }
+
+    states = statesBuilder.result()
+
+    initCleanupTimeState("QueryableStateCleanupTime")
+  }
+
+  override def processElement(
+    value: JTuple2[JBool, Row],
+    ctx: ProcessFunction[JTuple2[JBool, Row], Void]#Context,
+    out: Collector[Void]): Unit = {
+    if (value.f0) {
+      for (i <- nonKeyIndices.indices) {
+        states(i).update(value.f1.getField(nonKeyIndices(i)))
+      }
+
+      val currentTime = ctx.timerService().currentProcessingTime()
```

Comment:
Seems like this is registered to work with processing time domain only. I was wondering would it be easier to do ITCase if the table sink also works with event time, or having another piece of event time queryable table sink ?

Reply:
I think it's right to be able to process both time domain, and I've added.